src/Symfony/Component/Semaphore/Store/LockStore.php

```diff
@@ -0,0 +1,134 @@
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Semaphore\Store;
+
+use Symfony\Component\Lock\LockFactory;
+use Symfony\Component\Lock\LockInterface;
+use Symfony\Component\Semaphore\Exception\SemaphoreAcquiringException;
+use Symfony\Component\Semaphore\Exception\SemaphoreExpiredException;
+use Symfony\Component\Semaphore\Key;
+use Symfony\Component\Semaphore\PersistingStoreInterface;
+
+/**
+ * @author Alexander Schranz <alexander@sulu.io>
+ */
+class LockStore implements PersistingStoreInterface
+{
+    public function __construct(
+        private readonly LockFactory $lockFactory,
+    ) {
+    }
+
+    public function save(Key $key, float $ttlInSecond): void
+    {
+        $this->getLocks($key, $ttlInSecond, create: true);
+    }
+
+    public function delete(Key $key): void
+    {
+        foreach ($this->getLocks($key, 0, create: false) as $lock) {
+            $lock->release();
+        }
+        $key->setState(__CLASS__, null);
+    }
+
+    public function exists(Key $key): bool
+    {
+        $locks = $this->getLocks($key, 0, create: false);
+
+        if (\count($locks) !== $key->getWeight()) {
+            foreach ($locks as $lock) {
+                $lock->release();
+            }
+
+            return false;
+        }
+
+        return true;
+    }
+
+    public function putOffExpiration(Key $key, float $ttlInSecond): void
+    {
+        $locks = $this->getLocks($key, $ttlInSecond, create: false);
+        foreach ($locks as $lock) {
+            if ($lock->isExpired()) {
+                foreach ($locks as $lock) {
+                    $lock->release();
+                }
+
+                throw new SemaphoreExpiredException($key, 'One or multiple locks are already expired.');
+            }
+
+            $lock->refresh($ttlInSecond);
+        }
+
+        if (\count($locks) !== $key->getWeight()) {
+            foreach ($locks as $lock) {
+                $lock->release();
+            }
+
+            throw new SemaphoreExpiredException($key, 'One or multiple locks were not even acquired.');
+        }
+    }
+
+    /**
+     * @return array<LockInterface>
+     */
+    private function getLocks(Key $key, float $ttlInSecond, bool $create): array
+    {
+        $lockName = base64_encode($key->__toString());
+
+        $locks = [];
+        if (!$key->hasState(__CLASS__) && $create) {
+            $limit = $key->getLimit();
+
+            // use a random start point to have a higher chance to catch a free slot directly
+            $startPoint = rand(0, $limit - 1);
+
+            for ($i = 0; $i < $limit; ++$i) {
+                $index = ($startPoint + $i) % $limit;
+
+                $lock = $this->lockFactory->createLock($lockName.'_'.$index, $ttlInSecond);
+                if ($lock->acquire(blocking: false)) { // use lock if we can acquire it else try to catch next lock
```

Comment:
one could early break, when there is no chance of acquiring the next lock

```
if ($lock->acquire(blocking: false)) {
  // ...
} else {
  $acquired = \count($locks);
  $required = $key->getWeight();
  $remaining = $key->getLimit() - $i;
  if ($acquired + $remaining < $required) {
    break; // no chance to achieve the goal
  }
}

```

Reply:
Good idea.

Reply:
@jderusse thx for the quick review 🙏